### PR TITLE
Dynamic share providers

### LINKED ISSUE #29215: Dynamic share providers / share types in Nextcloud
My colleague @ylebre is working on [a project](https://github.com/pondersource/nc-sciencemesh) that makes it possible to connect Nextcloud servers to [Sciencemesh](https://sciencemesh.io/).
For this project we need to add custom handling for [File Sharing](https://docs.nextcloud.com/server/latest/user_manual/en/files/sharing.html). This also means we need a custom sharing _type_.

For our custom install, we can just add another class constant to the [`lib/public/Share/IShare.php`](https://github.com/nextcloud/server/blob/HEAD/lib/public/Share/IShare.php) interface, and custom handling logic for this type.

But when we were looking at the code, we saw a comment in [the Share20 `Manager` class](lib/private/Share20/Manager.php) that says:

> We can't handle other types yet

Hmmm... "yet"? :thinking:

And in the Share20 [`ProviderFactory`](https://github.com/nextcloud/server/blob/master/lib/private/Share20/ProviderFactory.php), the latest addition (for share type `IShare::TYPE_DECK`)
uses `$this->getProvider('deck')` instead of `$this->getDeckShareProvider()`.

This got us thinking that Nextcloud might benfit from an implementation that makes is possible to dynamically add share providers (types and logic). This could be very useful for Nextcloud plugins.

So my question is:

**Is there any interest in merging such functionality, if provided?** (Including any docs or tests deemed necessary)

From what we can see, we _think_ this would be a relatively small change.
Files that we can see that need to be touched are:

- [`apps/files_sharing/lib/Controller/ShareAPIController.php`](https://github.com/nextcloud/server/blob/master/apps/files_sharing/lib/Controller/ShareAPIController.php) and [related test](https://github.com/nextcloud/server/blob/master/apps/files_sharing/tests/Controller/ShareAPIControllerTest.php)
- [`lib/private/Share20/Manager.php`](https://github.com/nextcloud/server/blob/b585cf1ea1baeb7bbfaf22953e48bbd33e8381fa/lib/private/Share20/Manager.php) and [related test](https://github.com/nextcloud/server/blob/b585cf1ea1baeb7bbfaf22953e48bbd33e8381fa/tests/lib/Share20/ManagerTest.php)
- [`lib/private/Share20/ProviderFactory.php`](https://github.com/nextcloud/server/blob/master/lib/private/Share20/ProviderFactory.php) (no tests seem to be available?)

Is the information I've shared here complete and correct? Am I missing anything?
And if there's an interest, which steps would we need to take in order to make this happen?

- - -

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by michielbdejong**
(still reviewing what the diff looks like, will mark as 'ready for review' soon)

Fixes https://github.com/nextcloud/server/issues/29215

### CODE REVIEW

**Comment by michielbdejong** [comment]
I'll fix the failing unit tests and also add some more unit tests to cover the newly added code.

**Comment by michielbdejong** [comment]
Can someone approve running the workflows please?

**Comment by michielbdejong** [comment]
Thanks! Something new to fix. :)

**Comment by michielbdejong** [comment]
@labkode yes sorry, we're working on getting all the actions green now, and after that we will request a review from 2 committers. We hope to be on time with this for Nextcloud 25.

@navid-dada The [`Lint / php-cs check (pull_request)` action](https://github.com/nextcloud/server/runs/6522078901?check_suite_focus=true) is still failing, can you have a look?

The [`PHPUnit / php8.1-oci (pull_request)` action](https://github.com/nextcloud/server/runs/6522078010?check_suite_focus=true) is saying phpunit is not installed, which sounds like it's unrelated to this PR, does someone know?

**Comment by michielbdejong** [comment]
12 workflows awaiting approval

**Comment by michielbdejong** [comment]
Can someone approve running the workflows again please?

**Comment by michielbdejong** [comment]
What is the next step now? How can we find 2 reviewers for this PR?

**Comment by michielbdejong** [comment]
@blizzz can you help, or point us to the right person, maybe?

**Comment by labkode** [comment]
@blizzz @skjnldsv can you please kindly provide a review for this PR?

File: lib/public/Share/IShare.php [hunk 1]
```diff
@@ -116,6 +116,11 @@ interface IShare {
 	 */
 	public const TYPE_DECK_USER = 13;
 
+	/**
+	 * @since ?
+	 */
+	public const TYPE_SCIENCEMESH = 1000;
```
Comment (PVince81) [inline lib/public/Share/IShare.php:122]: it's not really a dynamic share type if app developers still need to add constants here

Comment (michielbdejong) [reply lib/public/Share/IShare.php:122]: Yeah, true! We should move this into https://github.com/nextcloud/server/pull/31571/files#diff-dbbe017dd357504abc442a6f1d0305166520ebf80353f42814b3f879a3e241bcR168 probably. And then if two apps are installed that try to claim the same number, it should complain that you cannot have both those apps installed at the same time.

**Comment by michielbdejong** [comment]
Would it help if, instead of trying to create the dynamic share type registry for you, I change this PR so that it just adds "yet another" else statement for share type ScienceMesh to the lists in various places?

**Comment by michielbdejong** [comment]
Replaced by https://github.com/nextcloud/server/pull/36228

**Pull request closed without merging**